Use AzureClusterIdentity when running ci e2e tests

nader-ziada opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind other
-->

**What this PR does / why we need it**:
- use AzureClusterIdentity in ci to make sure it working in all cases

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
- Update aad-pod-identity to v1.8.0 which upgrades CRDs from apiextensions/v1beta1 to apiextensions/v1
- Update calico version used by templates and tests
```

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e

nader-ziada commented:
> the `private cluster` test specifically is failing every time, looking into it

I believe the private cluster test is failing because of this issue here https://azure.github.io/aad-pod-identity/docs/troubleshooting/#token-requests-calls-fail-with-io-timeout

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e-windows

nader-ziada commented:
more info about issues with tests failing in this PR can be found in this slack thread https://kubernetes.slack.com/archives/CEX9HENG7/p1621362051002500

tldr; 
 - the nmi pod is having dns issues and can't authenticate when the management cluster is running on the Azure (not on a kind cluster) as in the case of the `private cluster` test. The dns issues happen when using `dnsPolicy=ClusterFirstWithHostNet` as recommended for any pod using `hostNetwork=true` 
 - The worker node has no issues connecting and the `nmi` pod works fine with `dnsPolicy=ClusterFirst`
 - but there seems to be an issue with either `how we configure the nmi pod`, or `the azure cluster vnet`

nader-ziada commented:
So the real issue here is kernel bug we get stuck on when using calico with vxlan

more details here: https://github.com/projectcalico/calico/issues/3145

nader-ziada commented:
I think we need to wait for the next release of calico that should include this fix here

https://github.com/projectcalico/felix/pull/2811

nader-ziada commented:
removed the identity ref from the private cluster template for now

nader-ziada commented:
@CecileRobertMichon ready for review

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e-windows

CecileRobertMichon commented on test/e2e/common.go:
```diff
@@ -51,6 +51,8 @@ const (
 	AzureNodeSubnetCidr      = "AZURE_NODE_SUBNET_CIDR"
 	MultiTenancyIdentityName = "MULTI_TENANCY_IDENTITY_NAME"
+	ClusterIdentityName      = "CLUSTER_IDENTITY_NAME"
+	VMSSHPort                = "VM_SSH_PORT"
```
are you using VMSSHPort anywhere? It was removed in another PR

nader-ziada replied:
looks like the rebase messed up something, will clean up

nader-ziada replied:
done

CecileRobertMichon reviewed (commented):
should we get rid of the `different SP identity` test since it's now equivalent to the others?

nader-ziada commented:
> should we get rid of the `different SP identity` test since it's now equivalent to the others?

yes, will remove that

fiunchinho commented:
I see that the release note for this PR is `None` but it's upgrading `aad-pod-identity` to `v1.8.0` which upgrades CRDs from apiextensions/v1beta1 to apiextensions/v1. Should this be included in the release note?

nader-ziada commented:
> I see that the release note for this PR is `None` but it's upgrading `aad-pod-identity` to `v1.8.0` which upgrades CRDs from apiextensions/v1beta1 to apiextensions/v1. Should this be included in the release note?

good point, it really should have been a different PR, but was done as part fixing this up here, will update release notes for now and if reviewers feel I should split it out in a different PR, I can do that

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e

CecileRobertMichon reviewed (commented):
/lgtm
/assign @devigned 

@nader-ziada do you have any thoughts on how we can encourage users to use AzureClusterIdentites for their v1alpha4 clusters? Should the default template example have it? Related to #1445

nader-ziada commented:
> @nader-ziada do you have any thoughts on how we can encourage users to use AzureClusterIdentites for their v1alpha4 clusters? Should the default template example have it? Related to #1445

- document it in the quick start guide
- add it to the example templates?

CecileRobertMichon commented:
> 1) document it in the quick start guide

might not be as easy as setting env vars for quick start but if we do 2) the quickstart experience should stay the same

> 2) add it to the example templates?

Yeah I think we should consider doing this

> 3) add some deprecation warning (not sure how)

this we can do in the controller logs and/or as a k8s event whenever credentials are set through manager environment variables

nader-ziada commented:
I'll just try again before I debug too much 

/test pull-cluster-api-provider-azure-e2e

nader-ziada commented:
the private cluster test is failing again because its getting stuck on creating the bastion host, 
```
I0622 21:57:02.456316       1 azurebastion.go:46] controllers/AzureCluster "msg"="creating bastion host" "azureCluster"="capz-e2e-jp3qx9" "cluster"="capz-e2e-jp3qx9" "namespace"="capz-e2e-tq69po" "bastion"="capz-e2e-jp3qx9-azure-bastion"
```
it eventually times out the private cluster is never created, 

still looking into it

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e

CecileRobertMichon commented:
/lgtm

@nader-ziada can you follow up on 

> document it in the quick start guide
might not be as easy as setting env vars for quick start but if we do 2) the quickstart experience should stay the same
add it to the example templates?
Yeah I think we should consider doing this
add some deprecation warning (not sure how)
this we can do in the controller logs and/or as a k8s event whenever credentials are set through manager environment variables

nader-ziada commented:
> might not be as easy as setting env vars for quick start but if we do 2) the quickstart experience should stay the same



> document it in the quick start guide
> might not be as easy as setting env vars for quick start but if we do 2) the quickstart experience should stay the same
> add it to the example templates?
> Yeah I think we should consider doing this
> add some deprecation warning (not sure how)
> this we can do in the controller logs and/or as a k8s event whenever credentials are set through manager environment variables

@CecileRobertMichon it was thinking to wait until we also can do that for the private clusters so we don't have any exceptions in the quick start guide, wdyt?

CecileRobertMichon commented:
/test ls

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-conformance-v1alpha4
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

CecileRobertMichon commented:
/approve
/hold 
feel free to unhold whenever we have confirmation that the conformance tests above are creating clusters without issue

nader-ziada commented:
/test pull-cluster-api-provider-azure-conformance-v1alpha4
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

nader-ziada commented:
/test pull-cluster-api-provider-azure-conformance-v1alpha4
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

nader-ziada commented:
/test pull-cluster-api-provider-azure-conformance-v1alpha4
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

nader-ziada commented:
/test pull-cluster-api-provider-azure-capi-e2e

nader-ziada commented:
/test pull-cluster-api-provider-azure-capi-e2e

shysank commented:
/test pull-cluster-api-provider-azure-capi-e2e

nader-ziada commented:
/test pull-cluster-api-provider-azure-capi-e2e

nader-ziada commented:
/hold cancel

@CecileRobertMichon PTAL

all the tests are using AzureClusterIdentity except for `SelfHosted` and `PrivateCluster`

nader-ziada commented:
/test pull-cluster-api-provider-azure-capi-e2e

nader-ziada commented:
rebased on capi v0.4.0

CecileRobertMichon commented on azure/scope/identity.go:
```diff
@@ -182,9 +181,9 @@ func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context, resourceMa
-				clusterctl.ClusterctlMoveLabelName: "true",
+				clusterv1.ClusterLabelName:                   clusterMeta.Name,
+				infrav1.ClusterLabelNamespace:                clusterMeta.Namespace,
+				"clusterctl.cluster.x-k8s.io/move-hierarchy": "true",
```
why more away from using a constant here?

nader-ziada replied:
I'll fix that

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-conformance-v1alpha4
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

CecileRobertMichon commented on test/e2e/azure_test.go:
```diff
@@ -145,6 +168,43 @@ var _ = Describe("Workload cluster creation", func() {
 			})
 		})
+
+		Context("Running the self-hosted spec", func() {
```
why not keep this in the capi_e2e suite like it was before? It's not really testing functionality that can be regressed by CAPZ PRs

nader-ziada replied:
because for the identity secret to be moved when the cluster pivots to a management cluster running on Azure, it has to be in the same namespace as the cluster and running the test spec from capi doesn't allow me to know the namespace

CecileRobertMichon replied:
right but could we define our own spec and run it as part of pull-cluster-api-provider-azure-capi-e2e ?

nader-ziada replied:
oh okay, just move the calling it in capi_e2e?

CecileRobertMichon replied:
yeah that's what I meant, sorry if it wasn't clear

nader-ziada replied:
no worries, that makes better sense, will change it

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-conformance-v1alpha4
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

nader-ziada commented:
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-conformance-v1alpha4
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

CecileRobertMichon commented:
will this affect scripts/ci-entrypoint.sh? want to make sure we don't break https://testgrid.k8s.io/provider-azure-master-signal jobs that rely on it

nader-ziada commented:
> will this affect scripts/ci-entrypoint.sh? want to make sure we don't break https://testgrid.k8s.io/provider-azure-master-signal jobs that rely on it

you are right, `hack/create-dev-cluster.sh` doesn't create the secret or set the related `AzureClusterIdentity` fields, so would probably fail

nader-ziada commented:
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-conformance-v1alpha4
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

nader-ziada commented:
@CecileRobertMichon I think the last two commits cover everything,   waiting for tests now🤞

nader-ziada commented:
/test pull-cluster-api-provider-azure-capi-e2e

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e-exp

CecileRobertMichon commented on hack/create-identity-secret.sh:
```diff
@@ -0,0 +1,35 @@
+export AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE="default"
+export AZURE_CLUSTER_IDENTITY_CLIENT_ID="${AZURE_CLIENT_ID}"
+
+kubectl create secret generic "${CLUSTER_IDENTITY_SECRET_NAME}" --from-literal=clientSecret="${AZURE_CLIENT_SECRET}"
```
how do we make it obvious that users need to do this on their own? Should we document the steps in getting-started.md and the quickstart?

nader-ziada replied:
I updated the docs in this PR #1477 which refers to this doc here: https://capz.sigs.k8s.io/topics/multitenancy.html

we would need to update the quick start but that's in capi, right?

CecileRobertMichon commented:
@nader-ziada did you test Tilt with these changes? I believe we also need to update the Tiltfile to create the secret on the management cluster

nader-ziada commented:
> @nader-ziada did you test Tilt with these changes? I believe we also need to update the Tiltfile to create the secret on the management cluster

I did not try tilt, will try it now

CecileRobertMichon commented on templates/test/ci/cluster-template-prow-multi-tenancy.yaml:
```diff
@@ -418,6 +418,18 @@ data:
                             type: string
                         type: object
                       type: array
+                    serviceLoadBalancerIPs:
```
can we completely get rid of the multi-tenancy flavor now that all flavors are using multi-tenancy?

nader-ziada replied:
I removed it from templates and removed the test, missed this one

CecileRobertMichon commented on templates/cluster-template.yaml:
```diff
@@ -193,3 +197,17 @@ spec:
+  namespace: default
+spec:
+  allowedNamespaces: {}
+  clientID: ${AZURE_CLUSTER_IDENTITY_CLIENT_ID}
```
would it make sense to keep this as `AZURE_CLIENT_ID` so existing setups (tilt, etc) can reuse the existing var?

nader-ziada replied:
let me try

CecileRobertMichon commented on Tiltfile:
```diff
@@ -169,6 +169,15 @@ def capz():
 
+def create_identity_secret():
+    #create secret for identity password
+    os.putenv('CLUSTER_IDENTITY_SECRET_NAME', 'cluster-identity-secret')
```
will we need to add the namespace variable to tilt-settings (if so needs to be added to tilt docs), or could we just default it / set it here?

nader-ziada replied:
its defaulting to `default` here currently, same as the script `hack/create-identity-secret.sh` 

for the e2e tests, its in the same namespace as the cluster

not sure what we need to change?

CecileRobertMichon replied:
it looks like the templates expect `AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE` to be set, should that be set here? Same question for `CLUSTER_IDENTITY_NAME`.

Basically, I'm wondering if I pull the changes from this PR and run `make tilt-up` and then click on one of the flavors to deploy it, will everything work as it did before? or are there any additional vars that need to be defined in my tilt-settings.json file now? It would be nice if everything kept working as before.

nader-ziada replied:
I think it should work the same, I added the setting of the namespace and identity name in the tiltfile similar to the script

nader-ziada replied:
> @nader-ziada tried out tilt locally, found two issues:
> 
> * if you run tilt up a second time it fails with `error: failed to create secret secrets "cluster-identity-secret" already exists`: we should delete the secret and recreate it every tilt up so it's idempotent (same thing we do with configmaps)
> * the secret env var is getting printed in tilt logs, any way we can avoid that?

where does it delete configmaps? can't find that

CecileRobertMichon replied:
here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/Tiltfile#L174

CecileRobertMichon commented:
@nader-ziada tried out tilt locally, found two issues:
- if you run tilt up a second time it fails with `error: failed to create secret secrets "cluster-identity-secret" already exists`: we should delete the secret and recreate it every tilt up so it's idempotent (same thing we do with configmaps)
- the secret env var is getting printed in tilt logs, any way we can avoid that?

CecileRobertMichon commented on hack/create-identity-secret.sh:
```diff
@@ -0,0 +1,34 @@
+set -o errexit
+set -o nounset
+set -o pipefail
+
```
`set +o xtrace`

we want to make sure the secret doesn't get printed

CecileRobertMichon commented:
found https://github.com/tilt-dev/tilt-extensions/tree/master/secret, not sure if it would be better to use that to create the secret (don't know if it logs the secret value the same way)

nader-ziada commented:
> @nader-ziada tried out tilt locally, found two issues:
> 
> * if you run tilt up a second time it fails with `error: failed to create secret secrets "cluster-identity-secret" already exists`: we should delete the secret and recreate it every tilt up so it's idempotent (same thing we do with configmaps)
> * the secret env var is getting printed in tilt logs, any way we can avoid that?

- added the delete statement similar to the configmaps
- added the quiet=true option to prevent printing

CecileRobertMichon commented on Tiltfile:
```diff
@@ -169,6 +169,20 @@ def capz():
+    os.putenv('CLUSTER_IDENTITY_NAME', 'cluster-identity')
+
+    substitutions = settings.get("kustomize_substitutions", {})
+    os.environ.update(substitutions)
```
what is this update doing?

devigned commented on hack/create-identity-secret.sh:
```diff
@@ -0,0 +1,34 @@
+export AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE="default"
+
+kubectl create secret generic "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" --from-literal=clientSecret="${AZURE_CLIENT_SECRET}"
+kubectl label secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" "clusterctl.cluster.x-k8s.io/move-hierarchy=true" --overwrite=true
```
New line

devigned commented on hack/gen-flavors.sh:
```diff
@@ -17,15 +17,15 @@
 set -o nounset
 set -o pipefail
-#set -o verbose
+# set -o verbose
```
delete?

devigned commented on templates/flavors/aks/kustomization.yaml:
```diff
@@ -3,3 +3,6 @@ kind: Kustomization
 - cluster-template.yaml
+- ../../azure-cluster-identity
+patchesStrategicMerge:
+- ../../azure-cluster-identity/managedazurecluster-identity-ref.yaml
```
newline

devigned commented on templates/flavors/system-assigned-identity/kustomization.yaml:
```diff
@@ -1,5 +1,6 @@
+  - ../base
+  - ../default/machine-deployment.yaml
 patchesStrategicMerge:
   - patches/system-assigned-identity.yaml
```
newline

devigned commented on templates/flavors/user-assigned-identity/kustomization.yaml:
```diff
@@ -1,5 +1,6 @@
+  - ../base
+  - ../default/machine-deployment.yaml
 patchesStrategicMerge:
   - patches/user-assigned-identity.yaml
```
newline

devigned commented on templates/flavors/windows/kustomization.yaml:
```diff
@@ -3,4 +3,8 @@ resources:
+
+patchesStrategicMerge:
+  - ../../azure-cluster-identity/azurecluster-identity-ref.yaml
   
```
weird newline

devigned commented on templates/test/ci/prow-ci-version/kustomization.yaml:
```diff
@@ -7,3 +7,12 @@ patchesStrategicMerge:
+patchesStrategicMerge:
+  - ../patches/control-plane-ci-version.yaml
+  - ../patches/controller-manager.yaml
+  - patches/machine-deployment-ci-version.yaml
```
newline

devigned commented on templates/test/ci/prow-identity-from-env/mhc.yaml:
```diff
@@ -0,0 +1,15 @@
+  unhealthyConditions:
+    - type: E2ENodeUnhealthy
+      status: "True"
+      timeout: 30s
```
newline

devigned reviewed (changes requested):
All nits.

CecileRobertMichon commented:
Do we want to keep all 7 commits in the final merge or are some of these squashable?

nader-ziada commented:
> Do we want to keep all 7 commits in the final merge or are some of these squashable?

I think they are for different parts, but I can squash if you prefer

CecileRobertMichon commented:
> I think they are for different parts, but I can squash if you prefer

nope, all good, just checking they are distinct

nader-ziada commented:
applied @devigned comments about newlines

devigned reviewed (approved):
/lgtm

CecileRobertMichon commented:
/lgtm

nader-ziada commented:
thanks for all the help with debugging this PR @CecileRobertMichon @devigned @mboersma

devigned commented:
@nader-ziada great work on this one! This one was a tough task. Thank you for taking this on.
